[GTK][WPE] Revamp focus rings

alice-mkh opened the pull request:
#### c63818df574ba14a0947d15860522e21cd3ffe31
<pre>
[GTK][WPE] Revamp focus rings
<a href="https://bugs.webkit.org/show_bug.cgi?id=241228">https://bugs.webkit.org/show_bug.cgi?id=241228</a>

Reviewed by Carlos Garcia Campos.

Use semi-transparent accent-colored focus rings, like macOS or GTK4.
Round them for radios and scales. Make sure to support CSS accent-color
for checks, radios and scales.

Use the same opacity as the libadwaita high contrast style and not the
regular one - the regular one doesn&apos;t work well on colorful backgrounds
like most webpages. This is also closer to the macOS style.

Make ThemeAdwaita::focusColor() private as it&apos;s both unused and doesn&apos;t
make much sense to be called from outside anymore.

This should fix the issue with modern media controls turning black
when focused.

* Source/WebCore/platform/adwaita/ThemeAdwaita.cpp:
(WebCore::ThemeAdwaita::focusColor):
(WebCore::getRectRadius):
(WebCore::ThemeAdwaita::paintFocus):
(WebCore::ThemeAdwaita::paintCheckbox):
(WebCore::ThemeAdwaita::paintRadio):
(WebCore::ThemeAdwaita::paintButton):
* Source/WebCore/platform/adwaita/ThemeAdwaita.h:
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::usePlatformFocusRingColorForOutlineStyleAuto):
(WebCore::useShrinkWrappedFocusRingForOutlineStyleAuto):
* Source/WebCore/rendering/RenderThemeAdwaita.cpp:
(WebCore::RenderThemeAdwaita::platformFocusRingColor const):
(WebCore::RenderThemeAdwaita::systemColor const):
(WebCore::RenderThemeAdwaita::paintTextField):
(WebCore::RenderThemeAdwaita::paintMenuList):
(WebCore::RenderThemeAdwaita::paintSliderTrack):

Canonical link: <a href="https://commits.webkit.org/252596@main">https://commits.webkit.org/252596@main</a>
</pre>

carlosgcampos commented on Source/WebCore/platform/adwaita/ThemeAdwaita.cpp:
```diff
@@ -81,29 +80,42 @@ Theme& Theme::singleton()
+    if (round)
+        radius = getRectRadius(rect, offset);
+    else
+        radius = 2;
```
```suggestion
    float radius = round ? getRectRadius(rect, offset) : 2;
```

carlosgcampos commented on Source/WebCore/platform/adwaita/ThemeAdwaita.cpp:
```diff
@@ -113,12 +125,17 @@ void ThemeAdwaita::paintFocus(GraphicsContext& graphicsContext, const Path& path
+        if (round)
+            radius = getRectRadius(rect, 0);
+        else
+            radius = 2;
```
```suggestion
        float radius = round ? getRectRadius(rect, 0) : 2;
```

carlosgcampos commented on Source/WebCore/platform/adwaita/ThemeAdwaita.h:
```diff
@@ -35,10 +35,9 @@ class Path;
+    static void paintFocus(GraphicsContext&, const FloatRect&, int offset, const Color&, bool round = false);
     static void paintFocus(GraphicsContext&, const Path&, const Color&);
-    static void paintFocus(GraphicsContext&, const Vector<FloatRect>&, const Color&);
+    static void paintFocus(GraphicsContext&, const Vector<FloatRect>&, const Color&, bool round = false);
```
I would use an enum here instead of a boolean, for example:

enum class PaintRounded : bool { No, Yes };

carlosgcampos reviewed (changes requested):
Looks good to me.

alice-mkh committed:
[GTK][WPE] Revamp focus rings
https://bugs.webkit.org/show_bug.cgi?id=241228

Reviewed by Carlos Garcia Campos.

Use semi-transparent accent-colored focus rings, like macOS or GTK4.
Round them for radios and scales. Make sure to support CSS accent-color
for checks, radios and scales.

Use the same opacity as the libadwaita high contrast style and not the
regular one - the regular one doesn't work well on colorful backgrounds
like most webpages. This is also closer to the macOS style.

Make ThemeAdwaita::focusColor() private as it's both unused and doesn't
make much sense to be called from outside anymore.

This should fix the issue with modern media controls turning black
when focused.

* Source/WebCore/platform/adwaita/ThemeAdwaita.cpp:
(WebCore::ThemeAdwaita::focusColor):
(WebCore::getRectRadius):
(WebCore::ThemeAdwaita::paintFocus):
(WebCore::ThemeAdwaita::paintCheckbox):
(WebCore::ThemeAdwaita::paintRadio):
(WebCore::ThemeAdwaita::paintButton):
* Source/WebCore/platform/adwaita/ThemeAdwaita.h:
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::usePlatformFocusRingColorForOutlineStyleAuto):
(WebCore::useShrinkWrappedFocusRingForOutlineStyleAuto):
* Source/WebCore/rendering/RenderThemeAdwaita.cpp:
(WebCore::RenderThemeAdwaita::platformFocusRingColor const):
(WebCore::RenderThemeAdwaita::systemColor const):
(WebCore::RenderThemeAdwaita::paintTextField):
(WebCore::RenderThemeAdwaita::paintMenuList):
(WebCore::RenderThemeAdwaita::paintSliderTrack):

Canonical link: https://commits.webkit.org/252596@main
